[Add] Function calling support for Autobuild

krishnashed opened the pull request:
## Why are these changes needed?

Introduction of Autobuild, didn't support function/tool calling. This PR is solution to add this feature/

## Related issue number

Closes #1510 

## Checks
Check and tests still need to be written and passed.

krishnashed commented:
@microsoft-github-policy-service agree

ekzhu commented:
You can fix code formatting running `pre-commit run --all-files` locally.

jamesliu commented:
@krishnashed, I'd like to explore and validate any test scenarios you might have. Thanks.

JieyuZ2 commented:
hey @krishnashed very nice feature, could you check this feature and make it compatible? https://microsoft.github.io/autogen/docs/topics/code-execution/user-defined-functions/

krishnashed commented:
Sure @JieyuZ2 I'll check it out and make sure it's compatible

krishnashed commented:
@JieyuZ2 @jamesliu I have added support for OpenAI Assistant generated using AutoBuild to access function calling. Also added working ipynb files for same

sonichi commented on OAI_CONFIG_LIST_sample:
```diff
@@ -21,4 +21,4 @@
         "api_version": "2024-02-15-preview"
     }
-]
+]
```
Remove this change.
```suggestion
]

```

krishnashed reviewed (commented):
Closing it as its implemented, and pre-build checks are now passing

krishnashed commented:
@jamesliu I have added test cases for Autobuild function calling, can you please review?

jamesliu commented on autogen/agentchat/contrib/agent_builder.py:
```diff
@@ -642,13 +686,58 @@ def _build_agents(
 
         Returns:
             agent_list: a list of agents.
+            list_of_functions: List of functions to be registered with agents.
```
list_of_functions in Args, not in Returns
